plugin/src/main/java/org/opensearch/ml/action/update/UpdateModelTransportAction.java

```diff
@@ -0,0 +1,564 @@
+        updateModelInput.setLastUpdateTime(Instant.now());
+        UpdateRequest updateModelGroupRequest = createUpdateModelGroupRequest(
+            newModelGroupSourceMap,
+            newModelGroupId,
+            newModelGroupResponse.getSeqNo(),
+            newModelGroupResponse.getPrimaryTerm(),
+            Integer.parseInt(updatedVersion)
+        );
+        try {
+            updateRequest.doc(updateModelInput.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS));
+            updateRequest.docAsUpsert(true);
+            if (isUpdateModelCache) {
+                String[] targetNodeIds = getAllNodes();
+                MLUpdateModelCacheNodesRequest mlUpdateModelCacheNodesRequest = new MLUpdateModelCacheNodesRequest(
+                    targetNodeIds,
+                    modelId,
+                    isPredictorUpdate
+                );
+                client.update(updateModelGroupRequest, ActionListener.wrap(r -> {
+                    client
+                        .update(
+                            updateRequest,
+                            getUpdateResponseListenerWithUpdateModelCache(modelId, wrappedListener, mlUpdateModelCacheNodesRequest)
+                        );
+                }, e -> {
+                    log
+                        .error(
+                            "Failed to register ML model with model ID {} to the new model group with model group ID {}",
+                            modelId,
+                            newModelGroupId,
+                            e
+                        );
+                    wrappedListener.onFailure(e);
+                }));
+            } else {
+                client.update(updateModelGroupRequest, ActionListener.wrap(r -> {
+                    client.update(updateRequest, getUpdateResponseListener(modelId, wrappedListener));
+                }, e -> {
+                    log
+                        .error(
+                            "Failed to register ML model with model ID {} to the new model group with model group ID {}",
+                            modelId,
+                            newModelGroupId,
+                            e
+                        );
+                    wrappedListener.onFailure(e);
+                }));
+            }
+        } catch (IOException e) {
+            log.error("Failed to build update request.");
+            wrappedListener.onFailure(e);
+        }
+    }
+
+    private ActionListener<UpdateResponse> getUpdateResponseListenerWithUpdateModelCache(
+        String modelId,
+        ActionListener<UpdateResponse> wrappedListener,
+        MLUpdateModelCacheNodesRequest mlUpdateModelCacheNodesRequest
+    ) {
+        return ActionListener.wrap(updateResponse -> {
+            // The update response returned an unexpected status may indicate a failed update
+            if (updateResponse != null && updateResponse.getResult() != DocWriteResponse.Result.UPDATED) {
```

Comment:
There will be 4 possible cases for this if clause, you just checked 1 edge case, but actually for the other 3 cases, only 1 case means success: `updateResponse != null && updateResponse.getResult() == DocWriteResponse.Result.UPDATED`

Reply:
Changed conditions